File: src/main/java/org/jetbrains/fortran/lang/parser/FortranManualPsiElementFactory.java
package org.jetbrains.fortran.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.fortran.lang.FortranTypes;
import org.jetbrains.fortran.lang.psi.impl.*;

public class FortranManualPsiElementFactory {
    public static PsiElement createElement(ASTNode node) {
        IElementType type = node.getElementType();
        if (type == FortranTypes.NUMERICAL_LABEL) {
            return new FortranNumericalLabelImpl(node);
        } else if (type == FortranTypes.NUMERICAL_LABEL_DECL) {
            return new FortranNumericalLabelDeclImpl(node);
        }else if (type == FortranTypes.LABELED_DO_CONSTRUCT) {
            return new FortranLabeledDoConstructImpl(node);
        } else if (type == FortranTypes.LABEL_DO_STMT) {
            return new FortranLabelDoStmtImpl(node);
        }

        return null;
    }
}

File: core/src/main/kotlin/org/jetbrains/fortran/lang/parser/FortranManualPsiElementFactory.kt
package org.jetbrains.fortran.lang.parser

import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import org.jetbrains.fortran.lang.FortranTypes
import org.jetbrains.fortran.lang.psi.impl.FortranLabelDeclImpl
import org.jetbrains.fortran.lang.psi.impl.FortranLabelDoStmtImpl
import org.jetbrains.fortran.lang.psi.impl.FortranLabelImpl
import org.jetbrains.fortran.lang.psi.impl.FortranLabeledDoConstructImpl

object FortranManualPsiElementFactory {
    @JvmStatic
    fun createElement(node: ASTNode): PsiElement? {
        val type = node.elementType
        return when {
            type === FortranTypes.LABEL                 -> FortranLabelImpl(node)
            type === FortranTypes.LABEL_DECL            -> FortranLabelDeclImpl(node)
            type === FortranTypes.LABELED_DO_CONSTRUCT  -> FortranLabeledDoConstructImpl(node)
            type === FortranTypes.LABEL_DO_STMT         -> FortranLabelDoStmtImpl(node)
            else -> null
        }
    }
}
